chess-AI/experimental-work/mcts.py

```diff
@@ -1,73 +1,161 @@
-        return nnet
+        """
+        NOTE: The search probabilites are being calculated based on the
+        the number of times a node has been visited/sum of all nodes visited
+        """
+
+        # Gets the list of all the moves
+        keys = list(self.n_values[fen_string].keys())
+
+        return keys,search_probs
+
+    # Main function to traverse the tree
+    def search(self,board,nnet):
+
+        # Need to get the fen string for the current state
+        fen_string = board.fen()
+
+        # Assigns reward if game is over
+        if board.is_game_over():
+            return 1 if board.outcome().winner else 0
+
+        # Checks to see if the node has been visited
+        if fen_string not in self.states_visited:
+
+            # Has to add the state to the list of visited nodes
+            self.states_visited.append(fen_string)
+
+            # Needs to get neural network state representation from fen_string
+            
+
+            # Gets the input from neural network
+            #input_state = get_cnn_input(fen_string)
+            value,policy = nnet.predict(board,None)
+
+            """
+            NOTE: Right now the state representation is not complete
+            and working with the neural network
+            The input is being passed into predict but not being used
+            """
+
+            # Need to update P with the state and the policy
+            self.p_values.update({fen_string:policy})
+
+            return -value
 
+        # Finds the legal moves from the current state
+        legal_moves = board.legal_moves
 
-    def executeEpisde(game,nnet):
+        # Finds the best move from the current state
+        move = self.find_best_move(legal_moves,fen_string)
 
-        examples = []
-        s = game.startState() # get the start state of the game
-        mcts = Mcts() # instantiate the MCTS class
- 
-        while True: 
+        # Makes the move on the board
+        board.push(move)
 
-            for _ in range(numMCTSims):
-                mcsts.search(s,game,nnet) #performs numMCTSims monte carlo simulations
+        # Search based on the new state
+        value = self.search(board,nnet)
 
-            examples.append([s,mcts.pi(s),None]) ## append the state, and improved policy, None refers to not knowing the values yet
+        # After search need to undo the move
+        board.pop()
 
-            a = random.choice(len(mcts.pi(s)), p=mcts.pi(s)) # choose a random move from the improved policy
-            s= game.nextState(s,a) # try that random move
+        # After undoing the move , the Q and N values need to be updated
+        fen_string = board.fen()
+        self.update_qn(fen_string,move,value)
 
-            if game.gameEnded(s): # if hte game is over then you need to assign rewards to all the examples
-                examples = assignRewards(examples,game.gameReward(s))
-                return examples
+        return -value
```

Comment:
Now that we understand why we are returning `-value`, you can add a comment to clarify.